mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/KeyRegistrationHandler.java

```diff
@@ -0,0 +1,132 @@
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.events;
+
+import static org.apache.james.backend.rabbitmq.Constants.AUTO_DELETE;
+import static org.apache.james.backend.rabbitmq.Constants.DURABLE;
+import static org.apache.james.backend.rabbitmq.Constants.EXCLUSIVE;
+import static org.apache.james.backend.rabbitmq.Constants.NO_ARGUMENTS;
+
+import java.nio.charset.StandardCharsets;
+
+import org.apache.james.event.json.EventSerializer;
+import org.apache.james.mailbox.Event;
+import org.apache.james.mailbox.MailboxListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.rabbitmq.client.AMQP;
+import com.rabbitmq.client.Connection;
+import com.rabbitmq.client.Delivery;
+
+import reactor.core.Disposable;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+import reactor.rabbitmq.QueueSpecification;
+import reactor.rabbitmq.RabbitFlux;
+import reactor.rabbitmq.Receiver;
+import reactor.rabbitmq.ReceiverOptions;
+import reactor.rabbitmq.Sender;
+
+class KeyRegistrationHandler {
+    private static final Logger LOGGER = LoggerFactory.getLogger(KeyRegistrationHandler.class);
+
+    private final MailboxListenerRegistry mailboxListenerRegistry;
+    private final EventSerializer eventSerializer;
+    private final Sender sender;
+    private final RoutingKeyConverter routingKeyConverter;
+    private final Receiver receiver;
+    private final RegistrationQueueName registrationQueue;
+    private final RegistrationBinder registrationBinder;
+    private Disposable receiverSubscriber;
+
+    KeyRegistrationHandler(EventSerializer eventSerializer, Sender sender, Mono<Connection> connectionMono, RoutingKeyConverter routingKeyConverter) {
+        this.eventSerializer = eventSerializer;
+        this.sender = sender;
+        this.routingKeyConverter = routingKeyConverter;
+        this.mailboxListenerRegistry = new MailboxListenerRegistry();
+        this.receiver = RabbitFlux.createReceiver(new ReceiverOptions().connectionMono(connectionMono));
+        this.registrationQueue = new RegistrationQueueName();
+        this.registrationBinder = new RegistrationBinder(sender, registrationQueue);
+    }
+
+    void start() {
+        sender.declareQueue(QueueSpecification.queue()
```

Comment:
Non durable  means a greater inconsistency window during the start process (as events received between queue binding & queue consumption will be lost)

The benefits of being non-durable is very certainly performance.

This seems to make sens for this feature - but for the sake of completness, is this trade-off documented?

Also, let's avoid implicit rabbitMQ config, and keep these parameter explicits as they are far from easy to choose. Furthermore, explicitly having this might defend us against driver changes.

Reply:
then you have to specify empty string name to the queue. 

I thought in the meeting about presenting rabbitmq eventbus yesterday, you expected it was non durable. its queue name is randomly, when you stop eventbus, and start again, I will create another queue for key registration with different name, you could not re use the old queue and events in that queue.
The second point is, you stop an eventbus, if I remember well, key registration will be wiped out, then all events will be lost anyway